text/0000-kafka-topics-per-tenant.md

```diff
@@ -0,0 +1,116 @@
+
+## Motivation
+FOLIO should allow production of messages to single Kafka topics related to an application concern. Without this change, there are scaling and cost implications as the number of tenants increases in a FOLIO cluster.
+
+- **Performance**: 
+Similar to a database not having a theoretical limit on the number of rows/tables, Kafka does not have a limit on partition count. A major difference is that a partition is "heavier" than a row/table. Partitions have to be rebalanced, replicated with corresponding open server files for each partition. Partitions have an ongoing administrative cost while online. Creating numerous partitions like they are database rows/tables will spell disaster for an under-provisioned Kafka installation. This is why cloud providers ensure to have limits on partition counts. Provisioning a Kafka installation to support such a high partition count will be wasteful for FOLIO hosting providers. <p>
+On the client side, buffers are created for each partition in memory. So a producer that interacts with very many thousands of partitions will consume more memory.
+- **Cost**: Kafka's use in FOLIO is still limited, more FOLIO modules will continue to employ event driven techniques that will cause an explosion of partitions. For hosting providers that self host Kafka, more brokers or broker capacity will be needed to service load generated by the high number of partitions. For hosting providers who utilize cloud managed Kafka, cloud providers typically charge per partition or set limits on partition count for a particular broker size. For example, Confluent Cloud charges $13/partition/year with a maximum of 4096 partitions for one cluster. AWS charges $3679/year for a broker size with 4 vCPU and 16GB Memory and a soft limit of 1000 partitions per broker.
+
+## Out of Scope
+- Enabling tenant data security within a Kafka installation. Tenant data security meaning "How do we prevent a tenant's data from being processed in place of another?", "How can we prevent a tenant module from seeing data from other tenants?"
+- Moving forward with the [Temporary Kafka Security Solution](https://wiki.folio.org/x/YYVFAw).
+- Countering module multi-versioning i.e. more than one version of a module is installed in FOLIO with tenants able to target specific versions of a module. **Module multi-versioning is supported by OKAPI but not for modules that interact directly via Kafka. This is an existing deficiency prior to this changes documented by this RFC.**
+- FOLIO modules that do not use folio-kafka-wrapper to interact with Kafka.
```

Comment:
I am hesitant to remove because I dont want to modify mod-pub-sub. Based on what i can see, modules who use mod-pub-sub is a closed loop, same with folio-kafka-wrapper. So they don't influence each other. about 60 % of kafka topics belonging to FOLIO are in the folio-kafka-wrapper loop. 

I will remove this line, the ideal move would be to migrate from mod-pub-sub to gain the benefits of extra safeguards added to folio-kafka-wrapper.

Reply:
I think this is a good topic to bring up to the TC for deliberation. I have reworded this document to be less folio-kafka-wrapper specific. A decision can be to modify mod-pubsub, migrate to folio-kafka-wrapper, or leave mod-pubsub as is.